### ISSUE TITLE
interrupt issues

### ISSUE THREAD

**Issue by orangecms** [ch32-rs/ch32-rs#3]
The `interrupt!` macro is currently not exported. I tried using the nightlies repo, and for the `ch32v3,` I tried adding this:
```rs
diff --git a/ch32v3/src/lib.rs b/ch32v3/src/lib.rs
index 7246166..58eef19 100644
--- a/ch32v3/src/lib.rs
+++ b/ch32v3/src/lib.rs
@@ -22,4 +22,5 @@ pub use self::generic::*;
 
 #[cfg(feature = "ch32v30x")]
 pub mod ch32v30x;
-
+#[cfg(feature = "ch32v30x")]
+pub use ch32v30x::interrupt;
```

Now I can use the macro. But I'm not seeing the handler invoked yet when I register one. I think I am always running into the `DefaultHandler`.
I can use the `DefaultHandler` when overriding it - though I have no clue how to clear the pending interrupt then, so the handler keeps running infinitely.

And I found that some IRQs are missing, namely those before the watchdog:
https://gitlab.com/eta-for-ci/ch32v307-gnumake/-/blob/945a34b9cf168c2dc8451b044c51f47053528d54/wch/hal_libraries/ch32v307_hal/include/ch32v30x.h#L29

```h
 /******  RISC-V Processor Exceptions Numbers *******************************************************/
  NonMaskableInt_IRQn         = 2,       /* 2 Non Maskable Interrupt                             */
  EXC_IRQn                    = 3,       /* 3 Exception Interrupt                                */
  Ecall_M_Mode_IRQn           = 5,       /* 5 Ecall M Mode Interrupt                             */
  Ecall_U_Mode_IRQn           = 8,       /* 8 Ecall U Mode Interrupt                             */
  Break_Point_IRQn            = 9,       /* 9 Break Point Interrupt                              */
  SysTicK_IRQn                = 12,      /* 12 System timer Interrupt                            */
  Software_IRQn               = 14,      /* 14 software Interrupt                                */

 /******  RISC-V specific Interrupt Numbers *********************************************************/
  WWDG_IRQn                   = 16,      /* Window WatchDog Interrupt 
```

I can confirm `SysTicK_IRQn` to be correct. I set up a simple counter, and it triggers (I am logging the IRQ number in `DefaultHandler`). See here:
https://github.com/orangecms/ch32v307-rust/commit/e06321779a2896ee01560db9546b27e4e7089435

**Comment by orangecms** [comment]
I have manually added the IRQ numbers here:
https://github.com/orangecms/ch32-rs-nightlies/commit/c2a821c58744f98b89071998f3fcfecd2f5ebe59

**Comment by andelf** [comment]
When a dedicated `rt` crate for ch32v MCUs added, this might be resolved.
🤔

**Comment by orangecms** [comment]
That sounds good, whatever works best :)

**Comment by Cydget** [comment]
I'm pretty new to rust let alone embedded rust. Is there a temporary workaround to this issue? I tried change the cargo.toml file to include the ch32v-rs-nightly build with @orangecms's method of modifying the lib.rs file to include pub use ch32v20x::interrupt; But I was not able to use the interrupt macro at all(I know clearing interrupts issue would still need to be overcome). I looked at @nosuz 's interrupt sample for his ch32v103-hal, but he made his own fork of riscv-rt called ch32v-rt which I'm pretty sure will only work with ch32v1 devices. My question is, is there a straightforward way to use interrupts currently, or do I just need to wait until someone releases another fork of risv-rt for all ch32vXXX devices? 

EDIT: after taking a deeper look at ch32-rt's differences from riscv-rt, my prior assumption that it would not work due to the generations being different has changed. I had thought that it wouldn't work because the ch32v1x devices are QingKeV3 architecture and the ch32v203 is QingKeV4. The both do not support supervisor mode. From what I can tell most of the changes made to riscv-rt were to remove all references to supervisor mode and it also included the interrupt macro I was potentially looking to use. I'm unsure why @nosuz removed all code related to supervisor mode instead of just using feature gates. Perhaps maintaining those gates would be more work. Regardless, I think I could make his solution work for ch32v2 devices, but it might not align with @andelf 's desired vision for how to support interrupts. I also still don't know if it clears the interrupts properly, i'm too inexperienced to determine based on the code without testing in hardware.
